### DISCUSSION TITLE
GUI Colors

### DISCUSSION THREAD

**Discussion by berensn** [vkbo/novelWriter#746] [Q&A]
Looking in the default_dark\theme.conf file, are the names of things under palette, like 'window', the name of QT objects that have RGB values assigned to them? If not, how/where are those keywords defined so that new ones could be created if I wanted?

**Comment by vkbo** [comment]
I haven't yet added the documentation for how to make new themes, but it's fairly straightforward. You just need to copy an existing theme and modify the main `.conf.` file. There's also a `.css` file for further customisation if needed.

You cannot add labels. They correspond to the colour roles for the Qt QPalette class. See https://doc.qt.io/qt-5/qpalette.html#ColorRole-enum

The labels are parsed (and thus defined) here: https://github.com/vkbo/novelWriter/blob/3ee07c7da70508986303ecddb3a0a008258c0bf1/nw/gui/theme.py#L297-L313

**Reply by berensn** [reply]
Gotcha. I've used PyQt in a previous project, but that was 2 years ago and mine is not as sophisticated as yours. 
My theming didn't use `.conf`, I was basically targeting `Qt Object Classes` directly in `css` ie:
`QLineEdit, QTextEdit{
   background: #242424;
   border: 1px solid #3b3b3b;
   border-radius: 2px;
   color: #999;
   height: 20px;
   selection-background-color: gray;
}`

The css files I saw in the theme folders were empty so I will have to experiment a bit and see what happens. 
This whole quest for the install directory and GUI colors was brought about with the dark theme edit window being white. 
<img width="1051" alt="Screen Shot 2021-04-12 at 9 28 49 AM" src="https://user-images.githubusercontent.com/7515745/114402894-47e99700-9b72-11eb-87de-da18df424537.png">
Yes thats a screenshot from MacOS. I have all the OSes. 
incidentally, to run on mac I had to cd to `~/.local/bin` and run `py novelWriter` (py is my python alias)

**Reply by vkbo** [reply]
The `.conf` files were just a quick way to allow themes. I've already received user contributed themes (the solarized ones). This is my first project using Qt. I've worked with Gtk before and used CSS for styling like you mention. The CSS files are empty now (didn't use to be) because the QPalette approach seems to catch pretty much everything consistently. I saw no reason to remove the CSS option in case someone wanted to further customise the GUI. There are some hard coded CSS rules in the code here and there to remove unwanted borders and such as well, but I've generally tried to avoid using CSS.

As for the screenshot above, the GUI theme and syntax theme are separate. You have a dark GUI theme with the default light syntax theme. The GUI elements of the editor also uses the syntax theme colours for consistency. Just select a dark syntax theme in Preferences, and flip the switch that prefers icons for dark backgrounds, which will make the icons on the status bar easier to see.

Edit: The reason why they are separate is that the default is to just use the OS theme colours, which may be light or dark, so the user need to be able to select matching syntax and icon themes.

**Reply by vkbo** [reply]
> incidentally, to run on mac I had to cd to `~/.local/bin` and run `py novelWriter` (py is my python alias)

Yeah, I've had other people mention this too, both on Mac and Linux. I think I added it to the documentation at some point that `~/.local/bin` needs to be added to `$PATH`.

I wish the Python people would come up with an easier way to deploy GUI applications. Python is an excellent platform for smaller GUI applications like this. I've sort of given up on trying to solve the deployment challenge on my own. The minimal zip files are a reflection of that as I just generate them from Linux with a script, calculate the checksum, and upload them. It's simple, and doesn't require too much of the user I hope.

**Reply by berensn** [reply]
Cool, the highlighting tab is where I found the syntax options.
And no, its not in the documentation. I added an alias to my `.bashrc` file `alias nw='py ~/.local/bin/novelWriter'` which would work for mac and linux, using pip I didn't have to do anything for windows, but if you did the self install you'd probably have to add to `$PATH`

Yeah, Python is really nice for small GUIs and apps, and there really should be a better way to package them.
Just found this about pip, you can [change the install directory](https://stackoverflow.com/questions/2915471/install-a-python-package-into-a-different-directory-using-pip)
Looks like `pip install --user novelWriter` is easiest, as it installs under the `$USER` path and then when run, searches this path so nothing needs to be added to `$PATH`
The next best looks like `pip install --prefix /custom_install_folder novelWriter` so you don't have duplicate dependencies or
`pip install --target /custom_install_folder novelWriter` if you don't care about duplicate dependencies.

**Reply by vkbo** [reply]
> And no, its not in the documentation.

It's mentioned in the "Getting Started" section that the install location may not be in PATH.

In any case, there is an option in the `setup.py` file to fill a standalone folder for Windows that bundles Python and dependencies using the `--target` approach. It can be further packaged as an installer, but I only used it for a couple of releases.

**Reply by berensn** [reply]
Ah crap, my bad. 

Gotcha, making an installer sounds like a definite pain in the ass. As it is, its not hard to install. 
Now that I know where to go to customize the theme, I can use that as an excuse to procrastinate on writing an actual novel ;)

**Reply by vkbo** [reply]
> ... I can use that as an excuse to procrastinate on writing an actual novel ;)

Well, that's pretty much how novelWriter came to exist in the first place!
